### PR TITLE
Hub content

### PR DESCRIPTION
**Pull request by elisa-lj11**
## Test Plan
1. Load in "DomainContent/Hub/models/pill.fbx"
2. Set this as the server script: "DomainContent/Hub/pill/spawnPills.js"
Set this as the client script: "DomainContent/Hub/pill/swallowFXPill.js"
3. In desktop mode, click on the pill
4. You should hear a swallow effect, a "space ambient" effect, and your whole visual field should change to a randomly selected effect (available effects: WIREFRAME, Depth, Albedo, Normal, Metallic, Emissive) and the pill should disappear
5. While the effect is activated, nothing should happen if you click the pill again
6. It should reset to normal in 10 seconds
7. In HMD mode, grab inside the pill and pull out the clone pill and bring it up to your face to swallow
8. You should hear a swallow effect, a "space ambient" effect, and your visual field should change and the pill should disappear like in desktop mode
9. If you grab another pill and eat it, you should see a new effect and hear a new swallow effect and "space ambient" effect
10. Your visual field should return to normal after 10 seconds after the initial pill was eaten

### To be added in future iterations

- an original script spawner
- made a wearable

### CODE REVIEW

File: DomainContent/Hub/spawnPills.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    var pillProperties; 
+    var spawnPillInterval;
+    
+    var PillSpawner = function(){
+
+    };
+
+    PillSpawner.prototype = {
+        preload: function(entityID) {
+            print("preload for spawnPills");
```
Comment (misslivirose) [inline DomainContent/Hub/spawnPills.js]: Remove print statements

File: DomainContent/Hub/spawnPills.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+                position: SPAWN_POSITION,
+                dynamic: true,
+                userData : "{\"grabbableKey\":{\"grabbable\":true}}"
+            };        
+            spawnPillInterval = Script.setInterval(function() {
+                Entities.addEntity(pillProperties); 
+            }, CHECK_INTERVAL);
+        },
+        unload: function() {
+            print("unload spawnPills");
```
Comment (misslivirose) [inline DomainContent/Hub/spawnPills.js]: Remove print statements

File: DomainContent/Hub/swallowFXPill.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+    Pill.prototype = {
+        playback: null,
+        isInactive: true,
+      
+        checkIfNearHead: function() {
+            var position = Entities.getEntityProperties(_entityID).position;
+            if (_this.isInactive) {
+                if (Vec3.distance(position, MyAvatar.getJointPosition("Head")) < WITHIN_10_CM) {
+                    print("swallowing pill");
```
Comment (misslivirose) [inline DomainContent/Hub/swallowFXPill.js]: Remove print statements

File: DomainContent/Hub/swallowFXPill.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    function Pill() {
+        return;
+    }
+
+    Pill.prototype = {
+        playback: null,
+        isInactive: true,
+      
+        checkIfNearHead: function() {
+            var position = Entities.getEntityProperties(_entityID).position;
```
Comment (misslivirose) [inline DomainContent/Hub/swallowFXPill.js]: Only retrieve properties required in getEntityProperties

File: DomainContent/Hub/swallowFXPill.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                    _this.playSwallowEffect(_this.avatarHeadPosition);
+                }
+            }
+        },
+
+        enableVisualEffects: function() {
+            var size = VISUAL_EFFECTS.length - 1;
+            var index = Math.round(Math.random() * size);
+            var effect = VISUAL_EFFECTS[index];
+            print("effect is: " + effect);
```
Comment (misslivirose) [inline DomainContent/Hub/swallowFXPill.js]: Remove print statements

File: DomainContent/Hub/swallowFXPill.js [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+            if (index === 0) {
+                Render.getConfig("RenderMainView").getConfig(effect.split(":")[1])[effect.split(":")[2]] = true;
+            } else {
+                Render.getConfig("RenderMainView").getConfig("DebugDeferredBuffer").enabled = true;
+                Render.getConfig("RenderMainView").getConfig("DebugDeferredBuffer").mode = effect;
+            }
+        },
+
+        playSwallowEffect: function(position) {
+            print("playing swallow effect");
```
Comment (misslivirose) [inline DomainContent/Hub/swallowFXPill.js]: Remove print statements

File: DomainContent/Hub/swallowFXPill.js [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+            var editJSON = {
+                visible: false,
+                lifetime: age + LIFETIME,
+                collisionless: true
+            };
+            Entities.editEntity(_entityID, editJSON);                   
+        },
+
+        mousePressOnEntity: function(entityID, mouseEvent) {
+            print("pill has been clicked");
```
Comment (misslivirose) [inline DomainContent/Hub/swallowFXPill.js]: Remove print statements

File: DomainContent/Hub/swallowFXPill.js [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+            if (_this.isInactive) {
+                Render.getConfig("RenderMainView").getConfig("DebugDeferredBuffer").size = {x: -1, y: -1, z: 1, w: 1};
+                _this.isInactive = false;
+                _this.playSwallowEffect(_this.avatarHeadPosition);
+            }
+        },
+
+        preload: function(entityID) {
+            _this = this;
+            print("loading new pill");
```
Comment (misslivirose) [inline DomainContent/Hub/swallowFXPill.js]: remove print statements

File: DomainContent/Hub/swallowFXPill.js [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+            Script.update.connect(this.checkIfNearHead);
+            _entityID = entityID;
+            Render.getConfig("RenderMainView").getConfig("DebugDeferredBuffer").size = {x: -1, y: -1, z: 1, w: 1};
+            this.playback = {volume: VOLUME, position: Entities.getEntityProperties(_entityID).position};
+        },
+
+        unload: function(entityID) {
+            print("unloading pill");
```
Comment (misslivirose) [inline DomainContent/Hub/swallowFXPill.js]: Remove print statements

File: DomainContent/Hub/pill/swallowFXPill.js [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+                visible: true,
+                lifetime: -1,
+                dynamic: true,
+                collisionless: true,
+                gravity: GRAVITY
+            };
+            Entities.editEntity(entityID, editJSON);
+        },
+
+        // FIXME: if someone is holding the pill and disconnects the pills don't disappear
```
Comment (misslivirose) [inline DomainContent/Hub/pill/swallowFXPill.js]: Remove commented out code

File: DomainContent/Hub/pill/swallowFXPill.js [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+            _this = this;
+            if (DEBUG) {
+                print("loading new pill");
+            }
+            Script.update.connect(_this.checkIfNearHead);
+            _entityID = entityID;
+            Render.getConfig("RenderMainView").getConfig("DebugDeferredBuffer").size = {x: -1, y: -1, z: 1, w: 1};
+            _this.swallowPlayback = {
+                volume: SWALLOW_VOLUME,
+                position: Entities.getEntityProperties(_entityID).position,
```
Comment (misslivirose) [inline DomainContent/Hub/pill/swallowFXPill.js]: Only get position from getEntityProperties

File: DomainContent/Hub/pill/spawnPills.js [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    var spawnPillInterval;
+    var pillBox = [];
+    
+    var PillSpawner = function(){
+        // nothing here
+    };
+
+    PillSpawner.prototype = {
+        preload: function(entityID) {
+            print("preload for spawnPills");
```
Comment (misslivirose) [inline DomainContent/Hub/pill/spawnPills.js]: Remove print statement or wrap in debug

File: DomainContent/Hub/pill/spawnPills.js [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+                    var pillID = Entities.addEntity(pillProperties);
+                    pillBox.push(pillID);
+                } else {
+                    pillBox.pop();
+                }
+                
+            }, CHECK_INTERVAL);
+        },
+        unload: function() {
+            print("unload spawnPills");
```
Comment (misslivirose) [inline DomainContent/Hub/pill/spawnPills.js]: Remove print statement or wrap in debug print

File: DomainContent/Hub/pill/swallowFXPill.js [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+//
+// All assets are under CC Attribution Non-Commerical
+// http://creativecommons.org/licenses/
+//
+
+/* globals Render */
+
+(function() {
+
+    /* fun trippy sphere bug
```
Comment (misslivirose) [inline DomainContent/Hub/pill/swallowFXPill.js]: Remove commented out code

File: DomainContent/Hub/pill/swallowFXPill.js [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+// All assets are under CC Attribution Non-Commerical
+// http://creativecommons.org/licenses/
+//
+
+/* globals Render */
+
+(function() {
+
+    var SWALLOW_SOUNDS_URLS = [
+        "http://www.pacdv.com/sounds/people_sound_effects/gulp-2.wav",
```
Comment (thoys) [inline DomainContent/Hub/pill/swallowFXPill.js]: Please download this audio-clip and put it relative like the swallow sounds

**Pull request merged**